Acdp 224 (#7)
Actindo-AG/Documentation: ActindoWorkFlow/UserInterface/01_WorkflowElements.md

[!!Workflows](ActindoWorkFlow)

# Workflow elements

![Workflow](/Assets/Screenshots/ActindoWorkFlow/Workflows/Workflow.png "[Workflow]")

> 

Author: ```suggestion
The *Workflows* module allows customers to model their own processes to match their companies structure.
```

> A workflow is represented with a number of fixed elements, each of them having specific characteristics. These elements are used to design the workflow, which can be edited as necessary according to customer needs. For detailed information, see [Manage workflows](/ActindoWorkFlow/Operation/01_ManageWorkflows.md).

Author: ```suggestion
Such a workflow consists of a number of fixed elements, each of them having specific characteristics. These elements are used to design the workflow, which can be edited as necessary according to customer needs. For detailed information, see [Manage workflows](/ActindoWorkFlow/Operation/01_ManageWorkflows.md).
```

> A workflow runs from left to right, from a start point to an end point, which can be in turn linked to another workflow. A workflow must contain at least one action between the start and the end point, which are special types of places. The actions are linked through places with each other, connecting the corresponding actions input and output ports. All these elements are described more in detail below.

Author: ```suggestion
A workflow runs from a fixed start point to a fixed end point, which can be in turn linked to another workflow. A workflow must contain at least one action between the start and the end point, which are special types of places. The actions are linked through places with each other, connecting the corresponding actions input and output ports. All these elements are described more in detail below.
```

> [Info] The *Workflows* module is linked to different system modules and plugins, such as *Invoicing*, *Email*, *OmniChannels*, *Tasks*, among others. The actions available to create workflows depend on the modules installed in the system.

[comment]: <> (Processes - The data input at the start point, for example a document, goes through the different workflow stages -places and actions- until it reaches the end point. The process is than completed.)

## Places

![Place](/Assets/Screenshots/ActindoWorkFlow/Workflows/Place.png "[Place]")

A place is a step in the workflow and is represented by circle, usually provided with inward and outward arcs. Places serve as link between two actions, joining an output port from the previous action to the corresponding input port of the following action. Places linking mandatory input/output ports are displayed by default and cannot be deleted. Additional places can be added by clicking on the ![Add](/Assets/Icons/Plus01.png "[Add]") (Add) button next to the optional input/output port. For detailed information about the types of input/output ports, see [Input and output ports](#input-and-output-ports).

> 

Author: Add information from gitlab:
- A place can hold 0-n tokens (in the workflow process)
- Every workflow starts with exactly one place and ends with exactly one place -> start point/place and end point/place
- A place always has a data type (sometime more, sometimes less specific) so you have a general idea what kind of data flows through it

When clicking a place, it becomes highlighted and the corresponding element settings are displayed in the settings side bar.

![Settings side bar](/Assets/Screenshots/ActindoWorkFlow/Workflows/SettingsSideBar01.png "[Settings side bar]")

The place settings side bar is located to the right of the workspace and allows to define the place settings. The following fields are displayed:

  - *Key*  
  Place key. This field is read-only. The key is required for API access and must be system wide unique.

  - *Data container*  
  Use this field to define the data type to be used in the place. If necessary, click the ![Delete](/Assets/Icons/Cross02.png "[Delete]") (Delete) button to display the whole list of available data types. Enter a search text for the desired data type. The list of actions is filtered for your search text as you type. Select the ![Radio button](/Assets/Icons/Radiobutton.png "[Radio button]") (Radio button) corresponding to the appropriate option.

## Actions

![Action](/Assets/Screenshots/ActindoWorkFlow/Workflows/AddAction.png "[Action]")

> Actions are the individual activities being executed in the workflow and are represented by a rectangle.

Author: Add info from gitlab:
Every API endpoint of the modules automatically becomes available as an action and can be used in any workflow - but you can also build your own actions independent from the api.

> Actions are the individual activities being executed in the workflow and are represented by a rectangle.

Author: Unterschied Action in workflow (in gitlab transition) und action in workflow process (in gitlab process action)

> All actions contain a certain, predefined number of input and output ports, which can be mandatory or optional. For detailed information about the types of input/output ports, see [Input and output ports](#input-and-output-ports).

Author: number of input and output ports between 1 and n

  - (Action label and name)  
  The action label is an action identification tag and can be defined by user. The action name is a system wide predefined denomination of the activity to be performed and is read-only.

  - *Key*  
   Action key. This field is read-only. The key is required for API access and must be system wide unique.

  - *Label*  
  Action label. Click this field to edit the action label. The action label is changed both in the settings side bar and in the diagram.

  - *Queue type*  
  Click the drop-down list to select the applicable queue type. For detailed information about queue types, see [Configure the queue types](/ActindoWorkFlow/Integration/01_ConfigureQueueTypes.md).

  - *Task event*  
  Click the drop-down list to select the applicable task event.

  > [Info] This option is linked to the *Tasks* module. It is only displayed if the *Tasks* module is installed and the user has the necessary rights. The *ActindoWorkflow Process Execution Failed* task is preconfigured by default. A list of all existing task events is displayed in the *Events* menu entry of the *Tasks* module.

  - *Configuration*  

>   Depending on the action, there might be configuration options, for instance, defining an email where a receipt or document must be sent. Editing the action configuration requires advanced programming knowledge.

Author: Add info from gitlab: 
An action may have configuration options, but most of them didn't have

  > [Info] Not all actions have configuration options. In this case, the *Configuration* section is not displayed.

  - *Static inputs*  
  All unlinked inputs are listed in this section. The static inputs determine the data type that can be input in a port. They are preconfigured but can be further defined by the user clicking the ![Add](/Assets/Icons/Plus01.png "[Add]") (Add) button. Only valid JSON values are accepted. When being edited, static inputs display a small arc without a place in front of them in the workflow diagram.

[comment]: <> (Unsure! RS Oli/Hannah)

> 

Author: Add *Arc* as an element

## Input and output ports

![Input and output ports](/Assets/Screenshots/ActindoWorkFlow/Workflows/NewPlace.png "[Input and output ports]")

All actions contain a certain, predefined number of input and output ports. They are defined with a descriptive name, such as origin, data or destination.

Input ports are positioned to the left of the action and output ports are positioned to the right of the action. Both are represented by a square.

Both input and output ports can be of two types:

- **Mandatory**  
  Mandatory ports must be linked to a place, which in turn must be linked to an action. Places linked to a mandatory port cannot be deleted or removed. When adding a new action, mandatory input/output places are already displayed.

- **Optional**  
 Optional ports are displayed with a ![Add](/Assets/Icons/Plus01.png "[Add]") (Add) button next to each of them. By clicking the ![Add](/Assets/Icons/Plus01.png "[Add]") (Add) button a new place is added. which can in turn be linked to a new action. For detailed information about adding new actions, see [Create an action](/ActindoWorkFlow/Operation/01_ManageWorkflows.md#create-an-action).

> 

Author: add info about static inputs:
-> Gitlab: Normally, all tokens for the transitions execution are taken from the places connected to its input ports. But sometimes you want to configure static values for certain inputs. This is possible using the *static inputs*: If a transition has configured a static input on one or more of its input ports, it will behave is if there was a place connected to that input port that always contains one token (containing whatever value is statically configured). You still need at least one "actual" place with an "actual" token in order to make that transition executable.

[!!Workflows](ActindoWorkFlow)

# Workflow elements

> *Workflows > Menu entry Workflows > Tab OVERVIEW > Select workflow > Select workflow version*

Author: ich denke, den Pfad brauchen wir hier nicht

![Workflow](/Assets/Screenshots/ActindoWorkFlow/Workflows/Workflow.png "[Workflow]")

A workflow is represented with a number of fixed elements, each of them having specific characteristics. These elements are used to design the workflow, which
can be edited as necessary according to customer needs. For detailed information, see [Manage workflows](#01_ManageWorkflows.md).

> A workflow runs from left to right, from a start point to an end point, which can be in turn linked to another workflow. The data input at the start point, for example a document, goes through the different workflow stages (places and actions) until it reaches the end point. The process is than completed.

Author: hier vielleicht nochmal umformulieren oder den zweiten Teil streichen, denn das gilt nicht für den workflow, sondern den (workflow) process

> > [Info] Data is provided from different system modules and plugins, such as *Invoicing*, *Email*, *OmniChannels*, *Tasks*, among others. The actions available to create workflows depend on the modules installed in the system.

Author: siehe oben

> 

Author: Kurzer Satz dazu, welche Elemente es im Workflow gibt und im folgenden näher beschrieben werden

## Places

![Place](/Assets/Screenshots/ActindoWorkFlow/Workflows/Place.png "[Place]")

> A place is a step in the workflow and is represented by circle, usually with inward and outward arrows.

Author: place = link/connection between two actions; special places = start point, end point; in a process, the data type is handled in the places; arrows = arcs

> Places serve as link between actions, joining an output port from the previous action to the corresponding input port of the following action. Places linking mandatory input/output ports are displayed per default and cannot be deleted. Additional places can be added by clicking on the ![Add](/Assets/Icons/Plus01.png "[Add]") (Add) button next to the optional input/output port. For detailed information about the types of input/output ports, see [Input and output ports](#input-and-output-port).

Author: -> #input-and-output-ports (add "s" at the end of ports)

When clicking a place, it becomes highlighted and the corresponding element settings are displayed in the settings side bar.

> ![Settings side bar](/Assets/Screenshots/ActindoWorkFlow/Workflows/SettingsSideBar01.png "[Settings side bar]")

Author: Screenshot abgeschnitten

> The place settings side bar is located to the right of the workspace and displays the following information:

Author: umformulieren -> in der settings side bar können die einstellungen für den place definiert werden

  - *Key*  

>   Element key. This information is read-only.

Author: element = place 
information = field
add: "The key is required for API access and must be system wide unique."

  - *Data container*  

>   Use this field to define the data type to be used in the place.

Author: Infos einschließen zum suchen nach datentyp, auswahl des datentyps durch auswählen des entsprechenden Radiobuttons

Reviewer: Frage an Oli: Wenn ein Radio button ausgewählt wird, wird es nicht automatisch übernommen und auf dem Feld Data container angezeigt. Fehlt noch ein Button (APPLY o.ä.)? Oder sollte es automatisch erfolgen? (Man muss offenbar woanders auf dem Editor klicken, um die Auswahl zu übernehmen.)

>     - ![Search](/Assets/Icons/Search.png "[Search]") (Search)  

Author: hier rausstreichen -> in user interface

  Enter a search text in the search bar. The data type list is filtered according to the search text entered.

>     - ![Delete](/Assets/Icons/Cross02.png "[Delete]") (Delete)  

Author: hier rausstreichen -> in user interface

  Click this element to clear the data displayed in the search bar. A list of all available data types is displayed.

>     - ![Radio button](/Assets/Icons/Radiobutton.png "[Radio button]") (Radio button)  

Author: kein eigener Punkt -> bei data container beschreiben

  Select the radio button corresponding to the appropriate option.

## Actions

![Action](/Assets/Screenshots/ActindoWorkFlow/Workflows/AddAction.png "[Action]")

> Actions, also known as transitions, are the individual tasks being executed in the workflow and are represented by a rectangle.

Author: transitions streichen

Reviewer: Activities? The WF-net transitions are assigned to tasks or activities, and places are assigned to the pre/post conditions... https://en.wikipedia.org/wiki/Petri_net

> Actions can contain any number of input and output ports, which can be mandatory or optional. For detailed information about the types of input/output ports, see [Input and output ports](#input-and-output-ports).

Author: all actions contain a certain, predefined number of input and output ports

 When adding a new action, all compatible input and output ports are included. Mandatory input ports are already displayed and linked to a place, whereas optional ones are displayed with a ![Add](/Assets/Icons/Plus01.png "[Add]") (Add) button.

>  All places linked to inputs and outputs ports must be provided with one or more pieces of data, also known as tokens, for the actions to be executed and the workflow to be completed. Otherwise, the workflow stops at the point where tokens are missing. Once an action is executed, one or more output data (tokens) are generated, which in turn become(s) input data (tokens) for the next action. For detailed information about tokens, see [Tokens](#to-be-determined).

Author: hier wieder aufpassen mit dem unterschied von process und workflow!

When clicking an action, it becomes highlighted and the corresponding element settings are displayed in the settings side bar.

![Settings side bar](/Assets/Screenshots/ActindoWorkFlow/Workflows/SettingsSideBar02.png "[Settings side bar]")

The settings side bar is located to the right of the workspace and displays the following information:

>   - (Action label and name)  

Author: anpassen -> keine description, sondern das (benutzerdefinierte) Label und der Name der Action

  The action label can be changed in the *Label* entry field. The action name provides a detailed description of the task to be performed and is read-only.

>   - *Key*  

Author: siehe oben

   Element key. This information is read-only.

  - *Label*  
  Action label. Click this field to edit the action label. The action label is changed both in the settings side bar and in the diagram.

  - *Queue type*  
  Click the drop-down list to select the applicable queue type. For detailed information about queue types, see [Configure the queue types](#to-be-completed).

  - *Task event*  
  Click the drop-down list to select the applicable task event.

>   > [Info] This option is linked to the *Tasks* module. It is only displayed if the *Tasks* module is installed and the user has the necessary rights. The *ActindoWorkflow Process Execution Failed* task is preconfigured per default. A list of all existing task events is displayed in the *Events* menu entry of the *Tasks* module.

Author: müssen uns mal absprechen, ob wir grundsätzlich "per default" oder "by default" schreiben...

Reviewer: by default

  - *Configuration*  

>   Depending on the action, there might be configuration options, for instance, defining an email where a receipt or document must be sent. Editing the action configuration requires advanced programming knowledge.

Author: Editing the action configuration requires advanced programming knowledge. -> depending on the configuration

>   > [Info] Not all actions have configuration options. In this case, the *Configuration* field is not displayed.

Author: *Configuration* section

  - *Static inputs*  

>   The optional inputs are displayed. They can be edited by clicking the ![Add](/Assets/Icons/Plus01.png "[Add]") (Add) button.

Author: hinzufügen: alle noch nicht verlinkten inputs werden unter den static inputs aufgelistet. Mindestens 1 input muss mit place verbunden sein und darf kein static input sein, damit im process eine action ausgeführt werden kann

Reviewer: Gehört diese Info nicht zu den Prozessen: Mindestens 1 input muss mit place verbunden sein und darf kein static input sein, damit im process eine action ausgeführt werden kann. 

Vorschlag: All unlinked inputs are listed in this section. The static inputs determine the data type that can be input. They are preconfigured but can be further configured by the user clicking the ![Add](/Assets/Icons/Plus01.png "[Add]") (Add) button. Only JSON values are accepted. When being edited, static inputs display a small arc without a place infront of them in the workflow diagram.

[comment]: <> (Unsure! Check Wissentransfer!)

[comment]: <> (The static inputs determine the data type that can be input. They are preconfigured but can be further configured by the user. Only JSON values are accepted.)

## Input and output ports

![Input and output ports](/Assets/Screenshots/ActindoWorkFlow/Workflows/InputOutputPorts.png "[Input and output ports]")

> Every action can contain any number of input and output ports. They can be defined with a descriptive name, such as reporter, title or event, or be simply marked as *in* (input) and *out* (output) followed by a consecutive number.

Author: certain/predefined number of input and output ports per action
-> satz über namen anpassen (out/in soll es ja nicht mehr geben...)

[comment]: <> (The plan is to provide all input and output ports with a name)

Input ports are positioned to the left of the action and output ports are positioned to the right of the action. Both are represented by a square.

Both input and output ports can be of two types:

- **Mandatory**  

>   Mandatory ports must be linked to a place, which in turn must be linked to an action. Otherwise, the workflow stops and shows an error. Places linked to a mandatory port cannot be deleted or removed. When adding a new action, mandatory input/output places are already displayed.

Author: ...the workflow fails (weiß nicht, ob der überhaupt startet)
start and end place ebenfalls mandatory

Reviewer: Frage für Oli? Satz weggenommen. Wenn er stimmt, gehört auch nicht hier, sondern bei Prozessen.

> 

Reviewer: It should actually be "process", right?: Otherwise, the **workflow** stops and shows an error.

- **Optional**  

>  Optional ports are displayed with a ![Add](/Assets/Icons/Plus01.png "[Add]") (Add) button next to each of them. By clicking the ![Add](/Assets/Icons/Plus01.png "[Add]") (Add) button a new action can be added. For detailed information about adding new actions, see [Link](#to-be-completed).

Author: By clicking the ![Add](/Assets/Icons/Plus01.png "[Add]") (Add) button a new action can be added -> stimmt, aber zuerst wird ein place hinzugefügt

> ## Tokens

Author: -> wie besprochen: müssen wir in ein anderes Dokument schieben oder dieses Dokument erweitern auf Workflow and Process elements

![Tokens](/Assets/Screenshots/ActindoWorkFlow/Workflows/Tokens.png "[Tokens]")

[comment]: <> (Token belongs in Processes)

> A token is a piece of data needed to complete an action, for example a document.

Author: token = piece of data oder data?

Reviewer: Ich meine, Oli hat es im Wissentranfer "beliebiges Datum" genannt, daher mein bewusst unbestimmtes "piece of data". ;) Haben wir nicht unterschiedliche Datentypen, z. B. also arbitrary, scalar value, array value...?

The *Processes* menu entry displays a read-only view of any process taking place in real time and a list of all actions being executed in the selected process. For detailed information about tracking a workflow process, see [Link](#to-be-completed).

> A process is started with one token at the start point, which will be processed in each action executed and, if necessary, turned into several tokens. It is therefore possible that a place contains several tokens, for example, several delivery notes issued for different parts of a same order. Each token will then be processed separately, that means, that the subsequent actions will be executed once for each token available.

Author: ..at least one token... aus einem token werden nicht zwei, aber eine Action kann durch ein Token ausgelöst werden und meherer Tokens ausgeben.

Reviewer: A process is started with **at least** one token at the start point,... ? 
Ich habe es so verstanden, dass ein Prozess **nur** mit einem Token anfängt, oder eben nicht?

The workflow process view displays the actions in different colours depending on their current status:

  - Green: done
  - Blue: active
  - White: waiting
  - Red: error

> 

Author: hier fehlen noch status und einige heißen anders:
Ready
In Progress
Error
Done
Dead
Suspended
Aborted

[comment]: <> (Colour coding changing here too?)

> Next to the status of each action, a number is displayed. This number refers to the number of tokens processed in this action.  

Author: Darstellung der tokens in einer action hängt auch von status der action ab

> If a token is missing at some point, for example, if a document has not been received, the action cannot be executed and the process stops at the point where data are missing. In this case, a black square is displayed on the place specifying the number of tokens waiting to be processed.

Author: nicht komplett richtig -> all unprocessed tokens werden mit dem schwarzen rechteck angezeigt, da muss nicht unbedingt ein fehler vorliegen. Z.B. auch der Fall, wenn eine Aktion zwei Tokens aus unterschiedlichen places benötigt und das token noch nicht in einem der places vorhanden ist, weil eine vorherige Aktion noch nicht ausgeführt wurde -> vgl. Wissentransfer Part 1 ab Min. 41

> The tokens can also be displayed in a column view in the *Tokens* tab and filtered according to their status:  

Author: kein status-filter bei tokens

  - Done (green)
  - Being processed (blue)
  - Ready (grey)
  - Error (red)

[comment]: <> (Colour coding changing here too?)

For detailed information, see [Link](#to-be-determined).

## Start and end point

![Start and end point](/Assets/Screenshots/ActindoWorkFlow/Workflows/Workflow.png "[Start and end point]")

> A workflow must have a start and an end point. A piece of data, or token, initiates the workflow at the start point and goes through all stages (places and actions) until it reaches the end point. The end point is the final result of all actions executed in the workflow. A new (sub)process can be linked subsequently to the end point. for detailed information, see [Link].

Author: siehe oben -> token = data
auf Unterschied workflow/ workflow process achten
start point/end point = places
final result = certain data

> ### Start point

Author: vgl. places

The start point is the initial stage of the workflow. It is represented by a circle with an outward arrow pointing towards an input port of the first action. Similarly to a place, when clicking it, the start point becomes highlighted and the corresponding element settings are displayed in the settings side bar.

![Start point](/Assets/Screenshots/ActindoWorkFlow/Workflows/StartPoint.png "[Start point]")

The settings side bar is located to the right of the workspace and displays the following information:

  - *Key*  
  Element key. This information is read-only.

  - *Data container*  
  Use this field to define the data type contained to be used in the place.

    - ![Search](/Assets/Icons/Search.png "[Search]") (Search)  
  Enter a search text in the search bar. The data type list is filtered according to the search text entered.

    - ![Delete](/Assets/Icons/Cross02.png "[Delete]") (Delete)  
  Click this element to clear the data displayed in the search bar. A list of all available data types is displayed.

    - ![Radio button](/Assets/Icons/Radiobutton.png "[Radio button]") (Radio button)  
  Select the radio button corresponding to the appropriate option.

> ### End point

Author: vgl. place
